recipes/full_finetune_distributed.py

```diff
@@ -103,6 +105,21 @@ def __init__(self, cfg: DictConfig) -> None:
                 "full fp16 training is not supported with this recipe. Please use bf16 or fp32 instead."
             )
 
+        if cfg.get("fsdp_cpu_offload", False):
+            if not cfg.optimizer.get("fused", False):
+                log.warning(
+                    """
+                    It is highly recommended to use fused optimizer implementation when CPU
+                    offloading, otherwise optimizer will use non vectorized kernels and cause massive slowdown.
+                    """
+                )
+            else:
+                # We are fusing + CPU offloading - check to make sure we're in a nightly.
+                if not utils.torch_version_ge("2.4.0"):
+                    raise RuntimeError(
+                        "Using fused optimizer on CPU is only supported in PyTorch nightly."
+                    )
+
```

Comment:
I'm not a fan of this if-else block being in the main recipe at all. This is starting to get back into the territory of the wild-wild-west recipe scripts that we spoke about a while back. Let's please brainstorm a bit about how to handle this and come up with the right design for this.

Reply:
I gave a suggestion in _recipe_registry.py, at least related to the versioning side of things (but it's a bigger change than what we're looking at here). Separately we are kinda drawing a line (I don't think?) we've drawn before.. we are putting an explicit nightly check in one of our core 4 recipes. While I 

As for the actual validation of fsdp_cpu_offload + optimizer.fused, I guess it's a bit tricky to put in a single util since it's a more cross-cutting config pair. The only alternative I can think of offhand is to move into a config validation, which I think we also had legitimate concerns about doing previously. (I guess we can't validate directly on the optimizer either since it gets instantiated after the FSDP wrapping)

Reply:
Versioning recipes make sense, though this portion has a bit more than versioning since I also want to warn if fused optimizer is not used with CPU offload. 

> we are putting an explicit nightly check in one of our core 4 recipes.

We are - but only under a very specific case - CPU offload with FSDP distributed training.

Can definitely add a helper function, though let me know if we should consider something else / alternative design. @ebsmothers @kartikayk